contracts/TopShotMarketV3.cdc

```diff
@@ -0,0 +1,346 @@
+                emit MomentWithdrawn(id: tokenID, owner: self.owner?.address)
+
+            // If not found in this SaleCollection, check V1
+            } else if let v1Market = self.marketV1Capability {
+                let v1MarketRef = v1Market.borrow()!
+
+                if v1MarketRef.getPrice(tokenID: tokenID) != nil {
+                    // withdraw the moment from the v1 collection
+                    let token <- v1MarketRef.withdraw(tokenID: tokenID)
+
+                    // borrow a reference to the main top shot collection
+                    let ownerCollectionRef = self.ownerCollection.borrow()
+                        ?? panic("Could not borrow owner collection reference")
+
+                    // deposit the withdrawn moment into the main collection
+                    ownerCollectionRef.deposit(token: <-token)
+                } else {
+                    panic("Token with the specified ID is not already for sale in either sale collection")
+                }
+
+            } else {
+                panic("Token with the specified ID is not already for sale in either sale collection")
+            }
+        }
+
+        /// purchase lets a user send tokens to purchase an NFT that is for sale
+        /// the purchased NFT is returned to the transaction context that called it
+        ///
+        /// Parameters: tokenID: the ID of the NFT to purchase
+        ///             buyTokens: the fungible tokens that are used to buy the NFT
+        ///
+        /// Returns: @TopShot.NFT: the purchased NFT
+        pub fun purchase(tokenID: UInt64, buyTokens: @FungibleToken.Vault): @TopShot.NFT {
+
+            // First check this sale collection for the NFT
+            if self.prices[tokenID] != nil {
+                assert(
+                    buyTokens.balance == self.prices[tokenID]!,
+                    message: "Not enough tokens to buy the NFT!"
+                )
+
+                // Read the price for the token
+                let price = self.prices[tokenID]!
+
+                // Set the price for the token to nil
+                self.prices[tokenID] = nil
+
+                // Take the cut of the tokens that the beneficiary gets from the sent tokens
+                let beneficiaryCut <- buyTokens.withdraw(amount: price*self.cutPercentage)
+
+                // Deposit it into the beneficiary's Vault
+                self.beneficiaryCapability.borrow()!
+                    .deposit(from: <-beneficiaryCut)
+                
+                // Deposit the remaining tokens into the owners vault
+                self.ownerCapability.borrow()!
+                    .deposit(from: <-buyTokens)
+
+                emit MomentPurchased(id: tokenID, price: price, seller: self.owner?.address)
+
+                // Return the purchased token
+                let boughtMoment <- self.ownerCollection.borrow()!.withdraw(withdrawID: tokenID) as! @TopShot.NFT
```

Comment:
Will there be any logic to cancel existing sales when a moment is transferred? Based on my novice understanding of cadence, I guess a sale would exist after transfer, and fail here when someone tries to make a purchase on a transferred moment?

Reply:
There isn't logic for it in the smart contract, but we could update the transfer moment transaction to also cancel the sale if it exists

Reply:
Do you think that you'd like to use a transaction like that?

Reply:
I think that could make sense - I'll defer to @sadief @EricLin2004 and @rrrkren here though.

Reply:
We don't borrow a capability to the moment itself, so nothing would happen to a capability because it doesn't exist. If the moment that was for sale is transferred, then when someone tries to purchase it, the purchase would fail and would revert. If we just put the cancel sale in the transfer transaction, then we would avoid this in the top shot marketplace

Reply:
We should talk about this. Thanks for clarifying on specifics but the fact remains that we will have some listings in the marketplace that are no longer valid if we don't remove them, yes?

Reply:
Can I clarify - The `cancelSale` already happens upon a transaction [here](https://github.com/dapperlabs/nba-smart-contracts/pull/118#discussion_r666547298) so confirming that there is no action required on this comment now? @EricLin2004 @joshuahannan

Reply:
Yes, that is correct